### PR TITLE
feat: Added preferred max loan parameter to leaser execute message

### LINKED ISSUE #32: [lease] opening a new lease should let the customer limit the borrowed amount
Currently, the process of opening a new lease borrows the maximum amount as configured with the initial liability percent. There is a requirement for providing the customer the flexibility to enter the lease with less liability than the pre-configured.

For example, if the init liability is set to 60% then for a downpayment of 100 USDC we provide a loan of 150 USDC, which is 60% of the total value of 100 + 150.

On opening a new lease, the customer should be able to provide an amount to borrow called `max_loan`. It is an optional argument. When provided, the actual amount to borrow should be equal to the min of `max_loan` and calculated per the initial liability configuration. If not, the existing algorithm applies.

For example, the customer may want to take only 100 USDC turning its lease into one with a 50% liability, or LTV ratio.

### PR DESCRIPTION
**Pull request by KirilMihaylov**
Closes #32

### CODE REVIEW

File: contracts/lease/src/api/open/mod.rs [hunk 1]
```diff
@@ -28,6 +31,9 @@ pub struct NewLeaseForm {
     pub customer: Addr,
     /// Ticker of the currency this lease will be about.
     pub currency: SymbolOwned,
+    /// Optional parameter indicating preferred maximum amount of the currency
+    /// the loan should be opened with.
+    pub max_loan: Option<CoinAmount>,
```
Comment (gmanev7) [inline contracts/lease/src/api/open/mod.rs:36]: After a thorough analysis of the requirements and the "ease-of-use" and "intuitive" interface principles, let's change the input to `max_ltv: Option<Percent>` with doc string:
`Maximum Loan-to-Value percentage of the new lease, optional`

File: contracts/lease/src/contract/cmd/open.rs [hunk 2]
```diff
@@ -61,18 +64,19 @@ impl<'a> WithLppLender for OpenLoanReq<'a> {
...
 
-            lpp.open_loan_req(borrow_lpn)?;
+        let borrow_lpn = self.liability.init_borrow_amount(downpayment_lpn);
 
-            Ok(Self::Output {
-                batch: lpp.into().batch,
-                downpayment,
-            })
-        }
+        lpp.open_loan_req(self.max_loan.map_or(borrow_lpn, Coin::new).min(borrow_lpn))?;
```
Comment (gmanev7) [inline contracts/lease/src/contract/cmd/open.rs:74]: perform the calculation as part of the `.init_borrow_amount(downpayment_lpn, max_ltv)`. 

Hint:
Apply `self.initial.max(max_ltv)` as a first step in the `fn Liability::init_borrow_amount`.

Update the current and add sufficient unit tests there.

File: contracts/leaser/src/contract.rs [hunk 3]
```diff
@@ -80,12 +80,13 @@ pub fn execute(
         ExecuteMsg::MigrateLeases { new_code_id } => owner_allowed_only(deps.storage, info, |s| {
             leaser::try_migrate_leases(s, new_code_id.u64())
         }),
-        ExecuteMsg::OpenLease { currency } => Borrow::with(
+        ExecuteMsg::OpenLease { currency, max_loan } => Borrow::with(
```
Comment (gmanev7) [inline contracts/leaser/src/contract.rs:83]: do propagate the changes in the Lease API to the Leaser API

File: packages/finance/src/liability/mod.rs [hunk 4]
```diff
@@ -91,18 +91,28 @@ impl Liability {
...
 
-        // borrow = init%.of(borrow + downpayment)
-        // (100% - init%).of(borrow) = init%.of(downpayment)
-        // borrow = init% / (100% - init%) * downpayment
-        let ratio = Rational::new(self.initial, Percent::HUNDRED - self.initial);
-        ratio.of(downpayment)
+        #[cfg(debug_assertions)]
+        if let Some(max_ltv) = max_ltv {
+            debug_assert!(max_ltv.percent() < Percent::HUNDRED);
+        }
```
Comment (gmanev7) [inline packages/finance/src/liability/mod.rs:106]: such debug assert makes sense on the value = min of the two ... it's more important to that... though self.initial is already check that is less than 100%

File: packages/finance/src/liability/mod.rs [hunk 5]
```diff
@@ -91,18 +91,28 @@ impl Liability {
...
+
+        let percent: Percent = max_ltv.map_or(self.initial, |percent: NonZeroPercent| {
```
Comment (gmanev7) [inline packages/finance/src/liability/mod.rs:108]: [repeating this comment again and again] names given after their type carry little info on what they are like. Please rename `percent` to `initial_ltv`

File: packages/finance/src/percent.rs
Comment (gmanev7) [inline packages/finance/src/percent.rs:1]: strongly against this new type. The benefits are negligible whereas the added noise to the code is significant.

File: contracts/leaser/src/cmd/quote.rs [hunk 6]
```diff
@@ -218,7 +218,7 @@ where
             return Err(ContractError::ZeroDownpayment {});
         }
 
-        let borrow = self.liability.init_borrow_amount(downpayment_lpn);
+        let borrow = self.liability.init_borrow_amount(downpayment_lpn, None);
```
Comment (gmanev7) [inline contracts/leaser/src/cmd/quote.rs:221]: this is not correct. 

we have to take `max_ltv` as an input quote parameter, otherwise the calculation would not be accurate

Comment (gmanev7) [reply contracts/leaser/src/cmd/quote.rs:222]: we have agreed to fix it

File: packages/finance/src/liability/mod.rs [hunk 7]
```diff
@@ -91,18 +91,27 @@ impl Liability {
...
 
-        // borrow = init%.of(borrow + downpayment)
-        // (100% - init%).of(borrow) = init%.of(downpayment)
-        // borrow = init% / (100% - init%) * downpayment
-        let ratio = Rational::new(self.initial, Percent::HUNDRED - self.initial);
-        ratio.of(downpayment)
+        #[cfg(debug_assertions)]
+        if let Some(max_ltv) = max_ltv {
+            debug_assert!(max_ltv > Percent::ZERO);
+        }
```
Comment (gmanev7) [inline packages/finance/src/liability/mod.rs:106]: agreed not to add artificial limits, remove it and add a test with `max_ltv = Zero`

**Comment by gmanev7** [comment]
LGTM
